### PR TITLE
Periodic donation

### PR DESCRIPTION
**Pull request by babygoat**

### CODE REVIEW

File: controllers/donation.go [hunk 1]
```diff
@@ -304,7 +402,35 @@ func buildPrimeSuccessRecord(resp tapPayTransactionResp) models.PayByPrimeDonati
...
+
+	m.CardInfoBinCode = &resp.CardInfo.BinCode
+	m.CardInfoLastFour = &resp.CardInfo.LastFour
+	m.CardInfoIssuer = &resp.CardInfo.Issuer
+	m.CardInfoFunding = &resp.CardInfo.Funding
+	m.CardInfoType = &resp.CardInfo.Type
+	m.CardInfoLevel = &resp.CardInfo.Level
+	m.CardInfoCountry = &resp.CardInfo.Country
+	m.CardInfoCountryCode = &resp.CardInfo.CountryCode
+	m.CardInfoExpiryDate = &resp.CardInfo.ExpiryDate
```
Comment (nickhsine) [inline controllers/donation.go]: Are those pointer values `nil` if tappay response does not contain `card_info`?

File: controllers/donation.go [hunk 2]
```diff
@@ -240,6 +322,22 @@ func buildClientResp(payMethod string, req tapPayPrimeReq, resp tapPayTransactio
...
 
+func buildDraftPeriodicDonation(userID uint, req clientReq) models.PeriodicDonation {
+	m := models.PeriodicDonation{}
+
+	m.CardholderEmail = req.Cardholder.Email
+	m.CardholderPhoneNumber = &req.Cardholder.PhoneNumber
+	m.CardholderName = &req.Cardholder.Name
+	m.CardholderZipCode = &req.Cardholder.ZipCode
+	m.CardholderAddress = &req.Cardholder.Address
+	m.CardholderNationalID = &req.Cardholder.NationalID
```
Comment (nickhsine) [inline controllers/donation.go]: Are those pointer values `nil` if `req` does not contain `card_holder`?

File: controllers/donation.go (hunk 1 above)
Comment (babygoat) [reply controllers/donation.go]: Nope. It will be empty string `""`

File: controllers/donation.go (hunk 2 above)
Comment (babygoat) [reply controllers/donation.go]: Nope. It will be empty string `""`

Comment (babygoat) [reply controllers/donation.go]: Currently, I cannot foresee any use case required to differentiate `NULL` from `""`.  I will replace all the fields with their underlying type.(i.e., remove pointer)

Comment (babygoat) [reply controllers/donation.go]: Currently, I cannot foresee any use case required to differentiate `NULL` from `""`. I will replace all the (*string) fields with their underlying type.(i.e., remove pointer)

File: models/donation.go [hunk 3]
```diff
@@ -96,20 +96,20 @@ type PeriodicDonation struct {
...
 	CardInfoFunding       *uint      `gorm:"type:tinyint" json:"card_info_funding"`
 	CardInfoType          *uint      `gorm:"type:tinyint" json:"card_info_type"`
-	CardInfoLevel         *string    `gorm:"type:varchar(10)" json:"card_info_level"`
-	CardInfoCountry       *string    `gorm:"type:varchar(30)" json:"card_info_country"`
-	CardInfoCountryCode   *string    `gorm:"type:varchar(10)" json:"card_info_country_code"`
-	CardInfoExpiryDate    *string    `gorm:"type:varchar(6);" json:"card_info_expiry_date"`
+	CardInfoLevel         string     `gorm:"type:varchar(10)" json:"card_info_level"`
+	CardInfoCountry       string     `gorm:"type:varchar(30)" json:"card_info_country"`
+	CardInfoCountryCode   string     `gorm:"type:varchar(10)" json:"card_info_country_code"`
+	CardInfoExpiryDate    string     `gorm:"type:varchar(6);" json:"card_info_expiry_date"`
```
Comment (nickhsine) [inline models/donation.go]: This change is kind of confusing.
So, if the tappay response is not containing `card_info`, will we store empty string `''` for `string` fields, and `nil` for `*uint` fields?

Comment (babygoat) [reply models/donation.go]: Yup, as I cannot see any difference between `null` and empty string `""`. Though, the raw record might be kind of tricky.

Comment (babygoat) [reply models/donation.go]: Fixed.

**Pull request merged**